File: digitalroot.rb
# http://rosettacode.org/wiki/Digital_root

def sum_digits(n,counter=0)
  n < 10 ? n + counter : sum_digits(n/10,counter+(n%10))
end

def find_root(n,persistence=1)
  n < 10 ? [n,persistence] : find_root(sum_digits(n),persistence+1)
end

[627615, 39390, 588225, 393900588225].each { |t| puts "#{t}: root = #{find_root(t)[0]} // persistence = #{find_root(t)[1]}" }

File: digital_root.c
// http://rosettacode.org/wiki/Digital_root

#include <stdio.h>

int sumDigits(int n, int counter)
{
    if (n < 10) {
        return n + counter;
    } else {
        counter += (n % 10);
        n /= 10;
        return sumDigits(n, counter);
    }
}

int findRoot(int number)
{
    if (number < 10) {
        return number;
    } else {
        number = sumDigits(number, 0);
        return findRoot(number);
    }
}

int main(int argc, const char * argv[]) {
    int value = findRoot(961038);
    printf("%d\n", value);
    return 0;
}
